src/lib/tls/tls_callbacks.h

```diff
@@ -255,6 +255,98 @@ class BOTAN_PUBLIC_API(2, 0) Callbacks {
                                       const std::vector<uint8_t>& msg,
                                       const std::vector<uint8_t>& sig);
 
+      struct Encapsulation_Result {
+            std::vector<uint8_t> encapsulated_bytes;
+            secure_vector<uint8_t> shared_secret;
+      };
```

Comment:
I'd prefer this be a class with a constructor, private members, and accessors. Specifically then, given an `Encapsulation_Result`, one can know that we never eg modify the encapsulated bytes.

Reply:
At first we just wanted to use a `std::pair` but went for this simple struct to be able to name the members of the returned tuple in `tls_kem_encapsulate`.

Though, thinking about it, we could add a `PK_KEM_Encryptor::encrypt()` override that also uses this struct instead of the currently used out-params. Having this as a "full blown" class might be more worthwhile.

Maybe as a follow-up?

Reply:
Yeah having something like this be the return type from the basic KEM operations would be quite reasonable, the current KEM approach of having the user pass two references is a little confusing. Main concern would be this is public API and whatever we ships in 3.1.0 (July 10th) we're mostly stuck with. If you think you would have time to look at this before then, fine to merge as is

Reply:
See here for a suggestion to generalize this result struct: https://github.com/randombit/botan/pull/3611